store/src/java/com/zimbra/cs/mailbox/Mailbox.java

```diff
@@ -10696,6 +9884,269 @@ public void execute(Message msg, MessageCallbackContext ctxt) {
         }
     }
 
+    public class MailboxTransaction implements AutoCloseable {
+        private final MailboxLock lock;
+        private boolean success = false;
+
+        public MailboxTransaction(String caller, OperationContext octxt, MailboxLock lock) throws ServiceException {
+            this(caller, System.currentTimeMillis(), octxt, lock,null, null);
+        }
+
+        public MailboxTransaction(String caller, OperationContext octxt, MailboxLock lock, RedoableOp recorder) throws ServiceException {
+            this(caller, octxt == null ? System.currentTimeMillis() : octxt.getTimestamp(), octxt, lock, recorder, null);
+        }
+
+        public MailboxTransaction(String caller, OperationContext octxt, MailboxLock lock, RedoableOp recorder, DbConnection conn) throws ServiceException {
+            this(caller, octxt == null ? System.currentTimeMillis() : octxt.getTimestamp(), octxt, lock, recorder, conn);
+        }
+
+        public MailboxTransaction(String caller, long time, OperationContext octxt, MailboxLock lock, RedoableOp recorder, DbConnection conn) throws ServiceException {
+            this.lock = lock;
+
+            boolean write = this.lock.isWriteLock() || requiresWriteLock();
+            assert recorder == null || write;
+            assert !Thread.holdsLock(this) : "use MailboxLock";
```

Comment:
This one is hurting my brain a little. How should I change this?

Reply:
If you were inside a synchronized method on the Mailbox or a block like synchronized(mailbox) { ... } then the Thread would hold an intrinsic lock on the Mailbox object and the assertion would fail and tell you to use a MailboxLock.  With the refactor now it's looking at the MailboxTransaction object.  I think it should be:

```
assert !Thread.holdsLock(this) : "use MailboxLock rather than synchronized";
assert !Thread.holdsLock(Mailbox.this): "use MailboxLock rather than synchronized";
```